Unified Cocaine ACL Tooling

karitra opened the pull request:
## Unified Cocaine ACL Tooling (UNICAT) plugin

Service for unified ACL records altering.

Preliminary implementation, ready to review.

**TODO:**
- [x] storage backend
- [x] unicorn backend
- [x] unicorn value version support!
- [x] complete grant/revoke pair for storage, unicorn
- [ ] ensure correct Unicorn prefix access semantics
- [x] ~~find better place for common stuff: `metainfo_t`, `indentity_t`, `flags_t`~~ (separate PR)
- [x] ~~tooling integration~~ (separate PR in tools)
- [ ] ensure correct cid/uid access restriction for grant/revoke (unitests?)
- [x] ~~additional methods: create/view/reset~~ (separate PR) 
- [x] ~~unit tests~~ (separate PR)

karitra committed:
wip: unicat, first blood

karitra committed:
feat(storage): backend impl

karitra committed:
feat(unicorn): preliminary impl

karitra committed:
refactor: attempt to make backends async

 Make backends async by building bit cosy async-callbacks hell.
 Need more testing.

karitra committed:
feat: propagate io exceptions to client

karitra committed:
wip: enlarge pile of garbage

  Added:
    - unicorn scope management
    - unicorn versioned values support (yet ugly)
    - metainfo is stored as stab with strings as a keys, for interop with dynamic_t
    - permissions set/unset code refactoring

karitra commented:
Yet lacks functional tests and marginal cases testing, but @antmat & @3Hren PTAL.

karitra committed:
fix: race condition on exception in async callback

    Fix for race in async callback on exception, which leads
    to owned (by callback) backend pointer to be released
    within same thread as callback was run on
    (in authorization::unicorn::executer).

karitra commented:
Yet lacks functional tests and marginal cases testing, but @antmat & @3Hren PTAL.

**UPD2**: following issue hopefully solved in 68f6e52.
**UPD1**: oops, testing results for unicorn, please standby:

```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::t
hread_resource_error> >'
  what():  boost thread: trying joining itself: Resource deadlock avoided
```

karitra committed:
feat: async wait all

    All async chains signal about normal or exceptional completion
    via shared completion_state_t. On destruction completion_state_t gathers all
    exceptions and sends slice of them to client or sends valid response to
    in case of nornal completion.

    Also cache for api::authorization::{unicorn_t,storage_t} was implemented,
    to avoid a race in self thread join on exception in ::verify calback by
    expanding api::authorization life-time to service life-span.

karitra commented:
@antmat, @3Hren, pls, don't beat me up for such a huge pile of garbage, but it seems the only way to animate all those async stuff, hopefully it will be party refactored out later, so PTAL.

3Hren commented on unicat/include/cocaine/auth/metainfo.hpp:
```diff
@@ -0,0 +1,146 @@
+    convert(const dynamic_t& from) {
+        auto& tuple = from.as_array();
+        if (tuple.size() != 2) {
+            throw std::bad_cast();
```
It's really misleading exception type. Consider `std::invalid_argument` instead.

3Hren commented on unicat/include/cocaine/detail/forwards.hpp:
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
Avoid `detail` section, especially in plugins, move in `src` instead.

3Hren commented on unicat/src/backend/backend.hpp:
```diff
@@ -0,0 +1,104 @@
+// Used in unicorn backend reincarnation for scope management.
+struct scope_ctl_t {
+    auto attach_scope(api::unicorn_scope_ptr s) -> void {
+        scope = s;
```
What to do with previous scope?

3Hren commented on unicat/src/backend/unicorn.cpp:
```diff
@@ -0,0 +1,110 @@
+    backend(api::unicorn(options.ctx_ref, options.name))
+{
+    COCAINE_LOG_DEBUG(this->logger(), "unicat::unicorn backend started '{}'", this->get_options().name);
+    dbg("unicorn backend is alive");
```
Do we need these logs?

3Hren commented on unicat/src/backend/unicorn.cpp:
```diff
@@ -0,0 +1,110 @@
+unicorn_backend_t::~unicorn_backend_t()
+{
+    COCAINE_LOG_DEBUG(this->logger(), "unicat::unicorn backend detached '{}'", this->get_options().name);
+    dbg("unicorn backend is dead");
```
And this.

3Hren commented on unicat/src/completion.hpp:
```diff
@@ -0,0 +1,122 @@
+                if (comp.has_error()) {
+                    if (comp.has_error_code()) {
+                        errors.push_back(comp.make_access_error());
+                        continue;
```
else?

3Hren commented on unicat/src/metainfo.cpp:
```diff
@@ -0,0 +1,150 @@
+#include "cocaine/auth/metainfo.hpp"
+#include "cocaine/idl/unicat.hpp"
+
+#if 0
```
Drop this doubtful stuff.

3Hren commented on unicat/src/module.cpp:
```diff
@@ -0,0 +1,14 @@
+
+extern "C" {
+    auto validation() -> cocaine::api::preconditions_t {
+        return cocaine::api::preconditions_t{COCAINE_MAKE_VERSION(0, 12, 0)};
```
12.14

3Hren commented on unicat/src/unicat.cpp:
```diff
@@ -0,0 +1,376 @@
+
+        for(const auto& el : entities) {
+            std::tie(scheme, service, entity) = el;
+            separated[cu::scheme_from_string(scheme)]
```
Simplify.

3Hren commented:
At last:

```
Undefined symbols for architecture x86_64:
  "_msgpack_unpack", referenced from:
      void cocaine::api::storage_t::get<cocaine::auth::metainfo_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<void (std::__1::future<cocaine::auth::metainfo_t>)>)::'lambda'(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >)::operator()(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >) const in storage.cpp.o
  "_msgpack_zone_destroy", referenced from:
      void cocaine::api::storage_t::get<cocaine::auth::metainfo_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<void (std::__1::future<cocaine::auth::metainfo_t>)>)::'lambda'(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >)::operator()(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >) const in storage.cpp.o
  "_msgpack_zone_init", referenced from:
      void cocaine::api::storage_t::get<cocaine::auth::metainfo_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<void (std::__1::future<cocaine::auth::metainfo_t>)>)::'lambda'(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >)::operator()(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >) const in storage.cpp.o
```

karitra committed:
fix: post review fixes

karitra commented:
@3Hren PTAL.

3Hren commented on unicat/src/unicat.cpp:
```diff
@@ -0,0 +1,357 @@
+
+namespace dbg {
+    auto
+    log_requested_entities(std::shared_ptr<logging::logger_t> log, const std::vector<cu::entity_type>& entities)
```
Seems like dead code.

karitra replied:
Yep, but it'll come handy in case of debug, should it be removed?

karitra committed:
refact: make storage ACL format backed by integer keys

karitra committed:
feat: log identity for every async opration

karitra committed:
refact: change order of access checks

    Change order of access checks and ACL metainfo alter operations:
        1. Check write access.
        2. Check read access.
        3. Read metainfo.
        4. Write metainfo.

karitra commented:
@3Hren, PTAL.

karitra committed:
refact: remove redundant code + update error messages

karitra committed:
fix: move share_ptr from referenced value

karitra commented:
Need bit more testing, probably unit-tests for metainfo alter cases + test for unicorn path, more intensive load test, exceptional cases test, but ready to be merged in order to be able to move further.
